feat: add adr-001 for node key refactoring (#608)
cosmos/iavl: docs/architecture/adr-001-node-key-refactoring.md

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.

The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes

## Decision

> - Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.

Reviewer: How is nodeKey computed, I remember it's sth like "version+path"?

Author: no, it is a just sequenced integer. 
for example, we assign the `nodeKey` of the new node as `tree.nonce + 1` every time when create node

Reviewer: It seems better to use "version/seq", where seq is only unique inside the version.

Author: here, the `nodeKey` is unique globally.

Reviewer: "version/seq" is globally unique as well, seq itself is local unique with the version. For each version, the nodes are written in a batch anyway, we only need to maintain the seq in memory.

Author: we keep `version => (root node key = first nonce, last nonce)`, and ensure the nodeKey is unique globally

Reviewer: I know, I was just suggesting to use "version/local-seq" as the nodeKey, so we don't need to store this nonce mapping at all.

Author: I know there are some advantages of `version|seq`, I am just worried it means we need to update `leftNodeKey` and `rightNodeKey` following this update and it leads to extra encode/decode executions and finally requires more storage to store the node data

Reviewer: If we constrain both version and seq to be 32bits, we can squeeze them into 8bytes :)

Reviewer: Another thing is "version|seq" can utilize dupsort feature of some db backends.

https://lmdb.readthedocs.io/en/release/

Reviewer 2: > I know there are some advantages of version|seq, I am just worried it means we need to update leftNodeKey and rightNodeKey following this update and it leads to extra encode/decode executions and finally requires more storage to store the node data

Why do we need to update the left and right node keys?

Author: here, `leftNodeKey` and `rightNodeKey` refer to the children node key in the db. If we update the node key as `version|seq`, we should also update left and right node keys, right?

Reviewer 2: Meaning that every time we have a new version we update all referenced node keys to refer to that version rather than the original version where they were created? That seems really complex. I feel like I'm maybe not understanding something really basic here

Author: @yihuang , I agree with you. If we keep the local nonce as `int32`, there is not much increase in storage size because we can remove the `version` from node body writes.

Reviewer: There are several places storage size would change, assuming version is `int64`:
- node key, 8bytes vs 8+4bytes, but with prefix compression, I think the real difference is minimal.
- node body, we can remove the version field, save a `varint(version)`.
- `leftNodeKey` and `rightNodeKey`, `varint(global nonce)` vs `varint(version << 32 + local nonce)`, there'll be a few more bytes here.

- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
- Remove the `orphans` from the tree.

Theoretically, we can also remove the `version` field in the node structure but it leads to breaking the ics23 proof mechanism. We will revisit it later.

New node structure

```go
type Node struct {
	key           []byte
	value         []byte
	hash          []byte    // keep this field in the storage
	nodeKey       int64     // new field, use as a node key

> 	leftNodeKey   int64     // new field, need to store in the storage

Reviewer 3: If we are using nonce as `nodeKey`, how are we getting data locality when traversing or writing the tree to disk?

I thought that the whole idea with the path encoded as a key was that writes would be sequential, contributing to lower frequency of compactions.

Reviewer: Without pruning, the nodes are all immutable, the `nodeKey` is sequential, so all the new nodes are written sequentially.

Author: I know the tree path describes the data locality well. The problem is how to keep the path when rotating the tree.

Reviewer: we need to split the `leftNodeKey` into two fields, `leftNodeVersion` and `leftNodeNonce`, right?

Reviewer: Aren't we need to split the `leftNodeKey` into `leftNodeVersion` and `leftNodeNonce`, likewise for `rightNodeKey`.

Author: right, I updated the node struct

	*ImmutableTree                                    
	lastSaved                *ImmutableTree
	nonce                    int64                    // new field to track the current ID
	versions                 map[int64]bool           
	allRootLoaded            bool                     
	unsavedFastNodeAdditions map[string]*fastnode.Node
	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

### Migration

We can migrate nodes one by one by iterating the version.

- Iterate the version in order, and get the root node for the specific version.
- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 

We will implement the `Import` functionality for the original version.

### Pruning

We introduce a new way to prune old versions.

For example, when a user wants to prune the previous 500 versions every 1000 blocks
- We assume the pruning is completed for `n`th version and the last nonce of `n`th version is `x`.
- We iterate the tree from the `n+501`th root node and pick only nodes which the nodeKey is in `[(n+1)th version first nonce, (n+500)th version the last nonce]`.

> - We can delete missing nodes instantly or re-assign the nodeKey from `x+1` in order for those nodes. Re-assign should be done after stopping the node but it can lead to improving the data locality.

Reviewer: I don't follow how pruning works, how do we know the node is not referenced by later versions so it's safe to delete?

Reviewer 2: I was wondering that as well, but not sure if I just don't understand

Author: we traverse all nodes based on `n+501`th root, and pick the nodes which the nodeKey is greater than `x`. 
then sort those nodes by nodeKey and can find the missing nodes.

for example, x = 500, nodes= [501, 502, 504, 506 ...], then we can assume 503 and 505 nodes are missed, right?

Reviewer: I see, use the n+501 tree as the gc root, and delete the missing ones. If a node is not referenced by n+501, it won't be referenced by later versions as well.

Author: exactly

Reviewer 2: I think this needs to be made a lot clearer and I'm still having trouble understanding how this prunes all the nodes. How do we know that nodes <= x are all pruned properly? Can't nodes less than x still be referenced by later versions of the tree and be eligible for pruning after the nth version has been pruned?

Also, generally this seems like a somewhat roundabout way to do pruning that could become inefficient with large trees because it requires traversing the whole tree and then storing nodes in memory in a sorted list to find out which nodes in the sequence are missing right?

Why don't we just do reference counting?

Reviewer 2: In my mind reference counting GC would make it simpler. It wouldn't require traversing all the nodes and would use very little memory. Has anyone explored that?

Author: The main idea of this story is to remove orphans, for me keeping orphans every time when updating the tree is very expensive. I'd like to separate pruning with main CRUD operations.

Reviewer 2: > I guess the cost of reference counting on disk is non-trivial though.

It's just an integer. I think we wouldn't know without benchmarks. Maybe we can explore it in the future.

Anyway, removing orphans makes sense.

Reviewer 4: traversing the whole tree is time-consuming. Shouldn't we try to avoid that?

Reviewer 5: pruning is handled from the application now, if we can keep the same pruning style, I understand there will need to be minor changes, then it would be good. Later on we should explore native pruning to avoid  write

## Consequences

### Positive

> Using the sequenced integer ID, we take advantage of data locality in the bTree and it leads to performance improvements. Also, it can reduce the node size in the storage.

Reviewer: Nowadays, all the "mainstream" db backends are lsm tree, but it should help there too.

> Using the sequenced integer ID, we take advantage of data locality in the bTree and it leads to performance improvements. Also, it can reduce the node size in the storage.

Author: OK, I will update this part and add more description of LSM

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

> The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.

Reviewer 6: Does the "data locality" make any significant difference in SSD disks?

Reviewer 6: Moreover, the data layout will depend on the underlying DB structure (goleveldb, rocksdb...) -- and most likely it will be different then we think. 
Do we have a benchmark proving the statement?

Author: https://github.com/cosmos/iavl/pull/597#issuecomment-1302490107
maybe it doesn't reflect fully the above decisions, it only refactored the nodeKey as a global nonce.

Reviewer 3: Data locality makes a difference in any LSM design.

The reason is that an LSM key-value store performs compactions under the hood to remove duplicated entries

Reviewer 7: Robert, data locality is the central problem for all computational work, including CPU caches, page tables and RAM.

Please read up on caches and the memory hierarchy.

Reviewer: I think "data locality" maybe a bit misleading here, because even with the new key format, we don't really have data locality for most of cases, because a node's children are likely belong to different versions, so they are still located distantly.
For me, the biggest advantage of new format is, the key is monotonically increasing, so it works better with kv-dbs where most of them store data in some kind of ordered trees, and our inserted keys are already ordered, and never modify(ignore pruning for now). We can even simply [dump the nodes into plain files](https://github.com/cosmos/iavl/pull/608#issuecomment-1301547758), they can still be queryable with some minimal metadata.

> The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:

Reviewer 6: ```suggestion
The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`. But there are only 2 use cases for `DeleteVersion`:
```

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes

## Decision

- Use the version and the sequenced integer ID as a node key like `bigendian(version) | bigendian(nonce)` format. 
- Remove the `version` field from node body writes.
- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
- Remove the `orphans` from the tree.

New node structure

```go
type NodeKey struct {
    version int64

>     nonce int32

Reviewer 3: I have a concern about the proposed nonce-approach still causing superfluous compactions and merging that are currently the biggest bottleneck here.

By writing key-value pairs that are already lexicographically sorted, we reduce superfluous merging and compactions overhead.

The main benefit of the initially proposed approach with a path encoded as a key is that writes would be sequential, reducing the superfluous merging.

While the "nonce approach" may show an improvement, it will still have redundant compactions due to nodes intra-version not being close to each other.

This might also affect iterations and queries.

Reviewer: With a monotonic increasing nonce, the inserted keys are also lexicographically sorted?

Reviewer: For example:
- version 1 create 3 nodes, they are inserted with keys: `(1, 0)`, `(1, 1)`, `(1, 2)`.
- version 2 create 2 nodes, they are inserted with keys: `(2, 0)`, `(2, 1)`
- etc.

Author: I agree with @yihuang, it is already sorted by the version, and also we will make the nonce sorted in `SaveBranch`.

}

type Node struct {
	key           []byte
	value         []byte
	hash          []byte    // keep it in the storage instead of leftHash and rightHash
	nodeKey       NodeKey   // new field, the key in the storage
	leftNodeKey   NodeKey   // new field, need to store in the storage
	rightNodeKey  NodeKey   // new field, need to store in the storage
	leftNode      *Node
	rightNode     *Node
    size          int64
	subtreeHeight int8
	persisted     bool
}
```

New tree structure

```go
type MutableTree struct {
	*ImmutableTree                                    
	lastSaved                *ImmutableTree
	versions                 map[int64]bool           
	allRootLoaded            bool                     
	unsavedFastNodeAdditions map[string]*fastnode.Node
	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

> We will restruct the nonce when save the current version. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.

Reviewer 6: What do you mean with "restruct"? How the nonce will be computed?

### Migration

We can migrate nodes one by one by iterating the version.

- Iterate the version in order, and get the root node for the specific version.
- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 

> 

Reviewer 6: 1. Shall we remove the old node?
2. I don't think we should update past versions. If "past" versions will be recalculated then valid proofs issued for a past version will not work any more. This could be an issue for IBC and relayers. cc: @ebuchman

Author: 1. yes, we will provide the pruning functionality, it is useful to remove the old unnecessary nodes and reduce the storage.
2. We won't update the `version` itself, and just assign the nonce.

We will implement the `Import` functionality for the original version.

> ### Pruning

Reviewer 6: this section needs more details.

We introduce a new way to prune old versions.

For example, when a user wants to prune the previous 500 versions every 1000 blocks

> - We iterate the tree based on the `n+501`th root and pick only nodes which the version is in `[(n+1), (n+500)]`.

Reviewer 6: so you want to iterate 500 trees? That will take lot of time.

> - We can delete missing nodes (which don't appear in the above traversing) instantly or re-assign the nonce in order for those nodes. Re-assign should be done after stopping the node but it can lead to improving the data locality.

Reviewer 6: how are you going to "instantly" remove missing nodes?

## Consequences

### Positive

Using the version and the sequenced integer ID, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes easy to find the key. Also, it can reduce the key and node size in the storage.

Removing orphans also provides performance improvements including memory and storage saving. Also, it makes it easy to rollback the tree. Because we can remove all nodes for which the `nodeKey` is greater than the specified version.

### Negative

> It requires extra storage to store the node because it should keep `leftNodeKey` and `rightNodeKey` to iterate the tree. Instead, we can delete`leftHash` and `rightHash` fields in the node and reduce the key size.

Reviewer 5: could we add how much extra per key would be added?

Author: it will reduce the node size as a result, but it's difficult to say how much because `EncodeVarint` is variant.
add `hash` +32byte
remove `leftHash`, `rightHash` -64byte
remove `version` max -8byte
add `leftNodeKey`, `rightNodeKey` max (8+4) * 2 = 24byte

so we can save 16byte in the node body

remove `hash` -32byte
add `version|nonce` 8 + 4 = 12 byte

so we can save 20byte in the node key

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on the disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.

The `orphans` are used to manage node removal in the current design and allow the deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`. But there are only 2 use cases for `DeleteVersion`:

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes

## Decision

- Use the version and the sequenced integer ID as a node key like `bigendian(version) | bigendian(nonce)` format. 
- Remove the `version` field from node body writes.
- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field in the node body.

> - Separate the `orphans` from the tree CRUD operations, and refactor the orphan store like `bigendian(to_version) | bigendian(from_version) | bigendian(nonce)`.

Reviewer: Should keep the `from_version` as prefix, because we need to iterate based on it, when doing pruning.

	leftNode      *Node
	rightNode     *Node
	subtreeHeight int8
}
```

New tree structure

```go
type MutableTree struct {
	*ImmutableTree                                     // The current, working tree.
	lastSaved                *ImmutableTree            // The most recently saved tree.
	unsavedFastNodeAdditions map[string]*fastnode.Node // FastNodes that have not yet been saved to disk
	unsavedFastNodeRemovals  map[string]interface{}    // FastNodes that have not yet been removed from disk
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool // If true, the tree will work like no fast storage and always not upgrade fast storage
	
	mtx sync.Mutex
}
```

We will assign the `nodeKey` when saving the current version in `SaveVersion`. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.

### Migration

We can migrate nodes through the following steps:

- Export the snapshot of the tree from the original version.
- Import the snapshot to the new version.
	- Track the nonce for the same version using int32 array of the version length.
	- Assign the `nodeKey` when saving the node.

### Pruning

> We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).

Reviewer 5: ```suggestion
The current pruning strategies allows for intermediate versions to exist. With the adoption of this ADR we are migrating to allowing only versions to exist between a range (50-100 instead of 1,25,50-100).
```

> Here we are introducing a new way how to get orphaned nodes which remove in the `n+1`th version updates.

Reviewer 5: not sure if orphans should be used here since above we mention removing them. can say historical versions

- Traverse the tree in-order way based on the root of `n+1`th version.
- If we visit the lower version node, pick the node and don't visit further deeply. Pay attention to the order of these nodes.
- Traverse the tree in-order way based on the root of `n`th version.
- Iterate the tree until meet the first node among the above nodes(stack) and delete all visited nodes so far from `n`th tree.
- Pop the first node from the stack and iterate again.

If we assume `1 to (n-1)` versions already been removed, when we want to remove the `n`th version, we can just remove the above orphaned nodes.

### Rollback

When we want to rollback to the specific version `n`

- Iterate the version from `n+1`.
- Traverse key-value through `traversePrefix` with `prefix=bigendian(version)`.
- Remove all iterated nodes.

## Consequences

### Positive

> Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.

Reviewer 5: ```suggestion
Using the version and a local nonce, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.
```

> Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.

Reviewer 5: ```suggestion
* Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.
```

```
# node body

add `hash`:							+32 byte
add `leftNodeKey`, `rightNodeKey`:	max (8 + 4) * 2 = 	+24 byte
remove `leftHash`, `rightHash`:			    		-64 byte
remove `version`: 					max	 -8 byte
------------------------------------------------------------
					total save	 	 16 byte

# node key

remove `hash`:			-32 byte
add `version|nonce`:		+12 byte
------------------------------------
		total save 	 20 byte
```

> Removing orphans also provides performance improvements including memory and storage saving.

Reviewer 5: ```suggestion
* Removing orphans also provides performance improvements including memory and storage saving.
```

### Negative

> The `Update` operation will require extra DB access because we need to take children to calculate the hash of updated nodes.

Reviewer 5: ```suggestion
*  `Update` operations will require extra DB access because we need to take children to calculate the hash of updated nodes.
```

> It doesn't require more access in other cases including `Set`, `Remove`, and `Proof`.

Reviewer 5: ```suggestion
  * It doesn't require more access in other cases including `Set`, `Remove`, and `Proof`.
```

> It is impossible to remove the individual version. The new design requires more restrict pruning strategies.

Reviewer 5: ```suggestion
* It is impossible to remove the individual version. The new design requires more restrict pruning strategies.
```

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on the disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.

The `orphans` are used to manage node removal in the current design and allow the deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`. But there are only 2 use cases for `DeleteVersion`:

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes

## Decision

- Use the version and the path as a node key like `bigendian(version) | byte array(path)` format. Here the `path` is a binary expression of the path from the root to the current node. 
	```
	`10101` : (right, left, right, left, right) -> [0x15]
	```

> - Store only the child node key for the below version in node body writes. Because it is possible to get the child path for the same version.

Reviewer 2: ```suggestion
- Store the child node key in the node body only when the child node is from an earlier version. For the same version, it is possible to guess the child node key and thus it doesn't need to be explicitly stored.
```

	```go
	func (node *Node) getLeftNode() (*Node, error) {
		if node.leftNode != nil {
			return node.leftNode, nil
		}
		if node.leftNodeKey != nil {
			return getNode(node.leftNodeKey) // get the node from the storage
		}
		return getNode(&NodeKey{
			version: node.nodeKey.version,
			path: node.nodeKey.path + '0',  // it will be more complicated in the real implementation

> 		})

Reviewer 2: Why are we getting a node with a new node key here? Doesn't this case mean that left is a leaf? I think we want to show how we're generating a new node key during insert/update

Author: this is just demonstrating how to skip `leftNodeKey` in node writes, without storing leftNodeKey, we can get the left node for the same version

Reviewer 2: I see

Reviewer 2: Does it make sense to at least store 1 bit of data as to whether the child is empty? This would prevent needing to do any sort of db read in that case

Author: what do you mean by the empty child? There are only two kind of nodes (inner node and leaf)
the `inner node` always have two child nodes

	}
	```
- Remove the `version` field from node body writes.
- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field in the node body.
- Remove the `orphans` completely from both tree and storage.

> New node structure

Reviewer 5: does this change the proofs of the tree?

If we assume `1 to (n-1)` versions already been removed, when we want to remove the `n`th version, we can just remove the above orphaned nodes.

### Rollback

When we want to rollback to the specific version `n`

- Iterate the version from `n+1`.
- Traverse key-value through `traversePrefix` with `prefix=bigendian(version)`.
- Remove all iterated nodes.

## Consequences

### Positive

Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.

```
# node body

add `hash`:					    +32 byte
add `leftNodeKey`, `rightNodeKey`:	max 8 + 8 = +16 byte
remove `leftHash`, `rightHash`:			    -64 byte
remove `version`: 			max	    -8	byte
------------------------------------------------------------
				total save	     24	byte

# node key

remove `hash`:				-32 byte
add `version|path`:			+16 byte
------------------------------------
			total save 	 16 byte
```

Removing orphans also provides performance improvements including memory and storage saving.

### Negative

> The `Update` operation will require extra DB access because we need to take children to calculate the hash of updated nodes.

Author: so we are getting `left` and `right` in Set and Remove but not splay in Update since we have `leftHash` and `rightHash` in the original version

> The `Update` operation will require extra DB access because we need to take children to calculate the hash of updated nodes.

Reviewer 5: the trade off here is storage saving on per node basis, right? maybe we mention this as a tradeoff

It doesn't require more access in other cases including `Set`, `Remove`, and `Proof`.

> It is impossible to remove the individual version. The new design requires more restrict pruning strategies.

Reviewer 5: can you elaborate on this?

Author: I think it is enough, it will remove `orphans` and it will be impossible to remove the intermediate version from storage.
And `pruning` part explains in more detail the new methods.

Reviewer: Even if removing orphaned, it's certainly possible to delete versions in between.

Reviewer 5: lets edit to include this.

	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

We will assign the `nodeKey` when saving the current version in `SaveVersion`. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.

### Migration

We can migrate nodes one by one by iterating the version.

- Iterate the version in order, and get the root node for the specific version.
- Iterate the tree based on the root and pick only nodes the node version is the same as the given version.
- Store them using the new node key format.

### Pruning

We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).

When we want to prune all versions up to the specific version `n`

- Iterate the tree based on the root of `n+1`th version.
- Iterate the node until visiting the node the version is below `fromVersion` and don't visit further deeply.
- Apply `DeletePath` for all visited nodes the version is below `n+1`.

```go
func DeletePath(nk *NodeKey) error {

> 	DeleteNode(node)

Reviewer: The node is part of n+1 tree, why delete?

Author: yeah, you are right, it will delete only the parents path.

	if nk.path is not root {
		DeletePath(&NodeKey{
			version: nk.version,

> 			path: parent(nk.path), // it looks like removing the last binary

Reviewer: It can only find the parent in the version the node is created in, but there could be multiple parents to delete in the range of versions to be deleted, this method don't delete all orphaned nodes.

Author: the multiple parents would be deleted when the same version child is deleted.

Reviewer: With the method you described, it'll only delete one parent of the node, right?

Author: why only the parent? the purpose of this function is to delete all parents of the path to the root.

Author: @yihuang , you are right. There is a case that has two lower version children when rebalancing the tree, for those nodes, they will are not removed in `DeletePath`. I introduced a new way to get orphaned nodes. This is a little tricky to implement but its idea itself is clear. Please focus on removing `orphans`. It will give us more flexibility.

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original node key of IAVL is a hash of the node and it does not take advantage of data locality on disk. The nodes are stored in a random location of the disk due to the random hash value, so it needs to do a random search of the disk to find the node.

The `orphans` are used to manage the removed nodes in the current version and allow to deletion of the removed nodes for the specific version from the disk through the `DeleteVersion`. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases, the first one is the rollback of the tree and the second one is to remove the unnecessary old nodes.

## Decision

- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
- Remove the `version` field from the node structure.
- Remove the `orphans` from the tree.

New node structure

```go
type Node struct {
	key           []byte
	value         []byte

> 	hash          []byte

Reviewer 5: are hash and key here the same value? a hash of the value?

Author: here, the `hash` is the hash value of node.

	leftHash      []byte
	rightHash     []byte
	nodeKey       int64     // new field, use as a key
	leftNodeKey   int64     // new field, need to store
	rightNodeKey  int64     // new field, need to store
	version       int64     // will remove
	size          int64
	leftNode      *Node
	rightNode     *Node
	subtreeHeight int8
	persisted     bool
}
```

New tree structure

```go
type MutableTree struct {
	*ImmutableTree                                    
	lastSaved                *ImmutableTree
	nonce                    int64                    // new field to track the current ID
	orphans                  map[int64]int64          // will remove
	versions                 map[int64]bool           
	allRootLoaded            bool                     
	unsavedFastNodeAdditions map[string]*fastnode.Node
	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

> ## Consequences

Reviewer 5: we should touch on migration paths here. I think this is the biggest question on how to upgrade existing nodes

Author: OK, added the migration part

### Positive

Using the sequenced integer ID, we take advantage of data locality in the bTree and it leads to performance improvements.

Removing orphans also provides performance improvements including memory and storage saving. Also, it makes it easy to rollback the tree. Because we will keep the sequenced segment IDs for the specific version, and we can remove all nodes for which the `nodeKey` is greater than the specified integer value.

### Negative

It requires extra storage to store the node because it should keep `leftNodeKey` and `rightNodeKey` to iterate the tree. Instead, we can delete the `version` in the node and reduce the key size.

> It can't delete the old nodes for the specific version due to removing orphans. But it makes `rollback` easier and it makes it possible to remove old nodes through `import` and `export` functionalities. The `export` will restruct the tree to make node IDs to a sequenced segment like (1 ... node_sieze).

Reviewer 5: could you elaborate on cant delete old nodes? how will this effect the current pruning behaviour used in the sdk. 

using import and export to remove old nodes seems like the wrong approach. 

if a user wants to prune every 500 versions every 1000 blocks it seems this would not be supported based on the above framing.

# ADR ADR-001: Node Key Refactoring

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original node key of IAVL is a hash of the node and it does not take advantage of data locality on disk. The nodes are stored in a random location of the disk due to the random hash value, so it needs to do a random search of the disk to find the node.

> The `orphans` are used to manage the removed nodes in the current version and allow to deletion of the removed nodes for the specific version from the disk through the `DeleteVersion`. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases, the first one is the rollback of the tree and the second one is to remove the unnecessary old nodes.

Reviewer 8: ```suggestion
The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes
```

## Decision

- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
- Remove the `version` field from the node structure.
- Remove the `orphans` from the tree.

New node structure

```go
type Node struct {
	key           []byte
	value         []byte
	hash          []byte    // keep this field in the storage
	leftHash      []byte    // will remove
	rightHash     []byte    // will remove
	nodeKey       int64     // new field, use as a node key
	leftNodeKey   int64     // new field, need to store in the storage

> 	rightNodeKey  int64     // new field, need to store in the storage

Author: `leftNode` and `rightNode` is only meaningful on memory side, we need these fields to get children from the storage side.

	version       int64     // will remove
	size          int64
	leftNode      *Node
	rightNode     *Node
	subtreeHeight int8
	persisted     bool
}
```

New tree structure

```go
type MutableTree struct {
	*ImmutableTree                                    
	lastSaved                *ImmutableTree
	nonce                    int64                    // new field to track the current ID
	orphans                  map[int64]int64          // will remove
	versions                 map[int64]bool           
	allRootLoaded            bool                     
	unsavedFastNodeAdditions map[string]*fastnode.Node
	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

> ## Consequences

Reviewer 5: nit: sorry consequences should come after migration and then we can add a pruning section talking about what you touched on below

### Migration

We can migrate nodes one by one by iterating the version.

> - Iterate the version in order, and get the root node for the specific version.

Reviewer 5: this may need a larger discussion but should it be lazily or stop the node migrate then restart? Any thoughts on that?

Author: Re-assign should be done after stopping the node. Maybe we can delete only nodes which not appear in iterating. It can be done without breaking off. I will introduce both ways.

Reviewer 5: I can grab some snapshots to test this migration and the time it would take. It's gonna be a long one.

Author: Sorry, I was confused with pruning!
I think lazy loading is possible but we should restart the node to switch database after loading, right?

Reviewer 5: hmm, maybe lazy migration is not possible. cause it needs to start from the lowest version in the db then move upwards, right?

Author: I think it is possible, but this should be done at the top level, not within iavl.
For example, we load it to the new IAVL lazily through `Export` of the original version and the new `Import` functionality, after syncing and block confirmation, we can switch the store to the new version.

Reviewer 5: yes that also works. not sure how that would work with go.mods though unless we call this v1 so we can import both?

Author: https://github.com/golang/go/wiki/Modules#version-selection
Maybe, we should create v2 of iavl.

- Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 

We will implement the `Import` functionality for the original version.

### Positive

Using the sequenced integer ID, we take advantage of data locality in the bTree and it leads to performance improvements. Also, it can reduce the node size in the storage.

Removing orphans also provides performance improvements including memory and storage saving. Also, it makes it easy to rollback the tree. Because we will keep the sequenced segment IDs for the specific version, and we can remove all nodes for which the `nodeKey` is greater than the specified integer value.

### Negative

It requires extra storage to store the node because it should keep `leftNodeKey` and `rightNodeKey` to iterate the tree. Instead, we can delete the `version`, `leftHash`, and `rightHash` fields in the node and reduce the key size.

It can't delete the old nodes for the specific version due to removing orphans. We introduce a new way to prune old versions.

> For example, when a user wants to prune the previous 500 versions every 1000 blocks

Author: @tac0turtle, I added the explanation of how to prune old nodes.

## Changelog

- 2022-10-31: First draft

## Status

Proposed

## Context

The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.

The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:

1. Rollback of the tree to a previous version
2. Remove unnecessary old nodes

## Decision

- Use the version and the sequenced integer ID as a node key like `bigendian(version) | bigendian(nonce)` format. 
- Remove the `version` field from node body writes.
- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
- Remove the `orphans` from the tree.

New node structure

```go
type Node struct {
	key           []byte
	value         []byte

> 	hash          []byte    // keep this field in the storage

Reviewer: I guess one consequence is proof generation is slower, previously we can get child hash directly, now we need to load the node?

Author: We are using the ics23 proof, which requires size, version, and height so I believe it doesn't lead to any delays.

	nonce         int64     // new field, the sequenced integer ID within the specific version
	leftNodeKey   int64     // new field, need to store in the storage
	rightNodeKey  int64     // new field, need to store in the storage
	version       int64     
	size          int64
	leftNode      *Node
	rightNode     *Node
	subtreeHeight int8
	persisted     bool
}
```

New tree structure

```go
type MutableTree struct {
	*ImmutableTree                                    
	lastSaved                *ImmutableTree
	nonce                    int64                    // new field to track the integer ID of the current version
	versions                 map[int64]bool           
	allRootLoaded            bool                     
	unsavedFastNodeAdditions map[string]*fastnode.Node
	unsavedFastNodeRemovals  map[string]interface{}   
	ndb                      *nodeDB
	skipFastStorageUpgrade   bool 

	mtx sync.Mutex
}
```

> ### Migration

Reviewer: An alternative way to migrate is using the state streamer, recreate the new iavl db with the state changes.

Reviewer 6: This is a breaking change, so validators need to update at once (at least 67% of them). So we need a migration process anyway.

Reviewer: I think it's not consensus breaking, since the root hashes don't change, validators don't need to update at the same time?

Reviewer: No, it can migrate IAVL nodes one by one, change references between them, but the hashes of each node is not changed.

Reviewer: The way node's hash computed is the same, node's relationship is not changed, the hashes of each node is not changed.

We can migrate nodes one by one by iterating the version.

- Iterate the version in order, and get the root node for the specific version.

> - Iterate the tree and assign the `nodeKey` to nodes which the node version equals. 

Reviewer: Migrate a archive node and pruned node will end up with different node keys, but it's probably ok.

Reviewer: Inconsistent node nonces could make the state-sync snapshot inconsistent as well?

Reviewer 5: On the global nonce they would be different but version | local nonce should be same, right?

Reviewer: Hmm, in migration, it depends on the order traversal, this is easy to specify. but when handling the insertion, how do we specify the logic, to allow alternative implementations be compatible with each other.

Author: I think the inconsistency of nodeKey is not a big problem, the only thing we need is to match the proof.

Reviewer 5: We can test this on a live network with state sync and a simple migration to see if it works to be sure

Reviewer: > I think the inconsistency of nodeKey is not a big problem, the only thing we need is to match the proof.

it depends on how we export the state sync snapshot I think, if we re-map the nodekey, then the stored one don't matter.

Reviewer: > Hmm, in migration, it depends on the traversal order, this is easy to specify, like preorder+ascending. but when handling the insertion, we should also specify the precise logic of nonce assignments, to allow alternative implementations be compatible with each other.

I think the nonce assignment should happens in the `SaveBranch`, which do preorder ascending traversal, pruning historical branches.

Author: you are right

When we want to prune all versions up to the specific version `n`

- Iterate the tree based on the root of `n+1`th version.
- Iterate the node until visiting the node the version is below `fromVersion` and don't visit further deeply.
- Apply `DeletePath` for all visited nodes the version is below `n+1`.

```go
func DeletePath(nk *NodeKey) error {
	DeleteNode(node)
	if nk.path is not root {
		DeletePath(&NodeKey{
			version: nk.version,
			path: parent(nk.path), // it looks like removing the last binary
		})
	}
}
```

### Rollback

When we want to rollback to the specific version `n`

- Iterate the version from `n+1`.
- Traverse key-value through `traversePrefix` with `prefix=bigendian(version)`.
- Remove all iterated nodes.

## Consequences

### Positive

Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.

```
# node body

add `hash`:											+32 byte

> add `leftNodeKey`, `rightNodeKey`:	max 8 + 8	=	+16 byte

Reviewer 4: Is it mean `bigendian(version) | byte array(path)` only uses 8 bytes?

We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).

Here we are introducing a new way how to get orphaned nodes which remove in the `n+1`th version updates.

- Traverse the tree in-order way based on the root of `n+1`th version.
- If we visit the lower version node, pick the node and don't visit further deeply. Pay attention to the order of these nodes.

> - Traverse the tree in-order way based on the root of `n`th version.

Reviewer: https://github.com/cosmos/iavl/issues/635
I proposed a method to prune version by traversing two version simutanously, it can be applied to current iavl tree, don't need the path as nonce.

Author: it looks similar with the above suggestion, I agree with you, the `getting orphans` actually is not required a `nonce` or `path`.
I think we can create a separate PR. @tac0turtle what about you?

Reviewer 5: makes sense.
